File: server/src/main/com/broll/gainea/server/core/goals/impl/all/G_KillAllMonsters.java
package com.broll.gainea.server.core.goals.impl.all;

import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.battle.BattleResult;
import com.broll.gainea.server.core.bot.strategy.GoalStrategy;
import com.broll.gainea.server.core.goals.Goal;
import com.broll.gainea.server.core.goals.GoalDifficulty;
import com.broll.gainea.server.core.map.Continent;
import com.broll.gainea.server.core.objects.Unit;
import com.broll.gainea.server.core.objects.monster.Monster;
import com.broll.gainea.server.core.player.Player;
import com.broll.gainea.server.core.utils.LocationUtils;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class G_KillAllMonsters extends Goal {

    private Continent continent;

    public G_KillAllMonsters() {
        super(GoalDifficulty.MEDIUM, "");
    }

    @Override
    public boolean init(GameContainer game, Player player) {
        List<Continent> continets = game.getMap().getAllContinents();
        Collections.shuffle(continets);
        for (Continent continent : continets) {
            this.continent = continent;
            int monsterCount = getMonsterCount();
            if (monsterCount >= 2) {
                this.difficulty = monsterCount >= 5 ? GoalDifficulty.MEDIUM : GoalDifficulty.EASY;
                this.setExpansionRestriction(continent.getExpansion().getType());
                this.text = "Auf dem Kontinent " + continent.getName() + " darf es keine Monster mehr geben";
                return super.init(game, player);
            }
        }
        return false;
    }

    private int getMonsterCount() {
        return (int) continent.getAreas().stream().flatMap(it -> LocationUtils.getMonsters(it).stream())
                .filter(Unit::isAlive).count();
    }

    @Override
    public void check() {
        if (getMonsterCount() == 0) {
            success();
        }
    }

    @Override
    public void killed(Unit unit, BattleResult throughBattle) {
        if (unit instanceof Monster && unit.getOwner() == null) {
            check();
        }
    }

    @Override
    public void botStrategy(GoalStrategy strategy) {
        strategy.setSpreadUnits(false);
        strategy.setPrepareStrategy(() -> {
            List<Monster> monsters = continent.getAreas().stream().flatMap(it -> LocationUtils.getMonsters(it).stream()).collect(Collectors.toList());
            int stars = monsters.stream().mapToInt(it -> it.getStars()).sum();
            strategy.updateTargets(monsters.stream().map(Monster::getLocation).collect(Collectors.toSet()));
            strategy.setRequiredUnits(stars);
        });
    }
}

File: server/src/main/com/broll/gainea/server/core/goals/impl/all/G_KillAllMonsters.kt
package com.broll.gainea.server.core.goals.impl.all

import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.battle.BattleResult
import com.broll.gainea.server.core.bot.strategy.GoalStrategy
import com.broll.gainea.server.core.goals.Goal
import com.broll.gainea.server.core.goals.GoalDifficulty.EASY
import com.broll.gainea.server.core.goals.GoalDifficulty.MEDIUM
import com.broll.gainea.server.core.map.Continent
import com.broll.gainea.server.core.objects.Unit
import com.broll.gainea.server.core.objects.monster.Monster
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.player.isNeutral
import com.broll.gainea.server.core.utils.getMonsters

class G_KillAllMonsters : Goal(MEDIUM, "") {

    init {
        libraryText = text("X")
        libraryDifficulties += EASY
        libraryDifficulties += MEDIUM
    }

    private lateinit var continent: Continent
    override fun init(game: Game, player: Player): Boolean {
        for (continent in game.map.allContinents.shuffled()) {
            this.continent = continent
            val monsterCount = monsterCount
            if (monsterCount >= 2) {
                difficulty = if (monsterCount >= 5) MEDIUM else EASY
                setExpansionRestriction(continent.expansion.type)
                text = text(continent.name)
                return super.init(game, player)
            }
        }
        return false
    }

    private fun text(continent: String) =
        "Auf dem Kontinet $continent darf es keine wilden Monster mehr geben"

    private val monsterCount: Int
        get() = continent.areas.flatMap { it.getMonsters() }
            .count { it.owner.isNeutral() && it.alive }

    override fun check() {
        if (monsterCount == 0) {
            success()
        }
    }

    override fun unitKilled(unit: Unit, throughBattle: BattleResult?) {
        if (unit is Monster && unit.owner.isNeutral()) {
            check()
        }
    }

    override fun botStrategy(strategy: GoalStrategy) {
        strategy.isSpreadUnits = false
        strategy.setPrepareStrategy {
            val monsters = continent.areas.flatMap { it.getMonsters() }
            val stars = monsters.sumOf { it.stars }
            strategy.updateTargets(monsters.map { it.location }.toSet())
            strategy.setRequiredUnits(stars)
        }
    }
}
